Manage pgAdmin users and preferences via CLI. #2483

khushboovashi opened the pull request:
**Features/Fixes:**
- Provided CLI Interface for user management. Detailed descriptions (including commands) can be found in the User Management documentation.
- Provided CLI Interface for Preferences. Detailed descriptions (including commands) can be found in the Preference documentation.
- Fixed an issue where Import Export servers through CLI were used to honour only Internal users.
- For PIP Python installation. We can run commands as shebang.
 **Shebang**: pgadmin4-cli get-prefs
 **With Python Interpreter**: /path/to/python /path/to/setup get-prefs

**Technical Aspects**
Used the typer package instead of argparse in setup.py. 
- Class-based CLI interface.
- Command changes ex:
-- Old command:  /path/to/python /path/to/setup.py --load-servers input_file.json
-- New command:  /path/to/python /path/to/setup.py load-servers input_file.json

**Affected Areas**
- Setup.py
- Import Export Servers (both CLI and UI)
- Manage Users
- Preferences

akshay-joshi reviewed (changes requested):
1) Fixed SonarQube issues in the 'setup.py' file.

**GUI Comments:**

1. Should we stop adding/deleting/updating user options in the Desktop Mode?
2. Remove the "--auth-source" option from the "add-user" command as it is of no use.
3. Remove auth-source as 'internal' from the "add-external-user" command, as we will unable to add internal users using this command. Add a validation if someone provides then the user should be warned.
4. Password validation of  6 characters should be there in the "add-user", "update-user" command.
5. Do we need the "confirm password" option in the "add-user" command?
6. Please check the already existing email or username if possible.
7. What is the difference/use between/of --console and --no-console? I have run both but have not found any difference, tried with add-user and add-external-user.
8. delete-user is a very risky option to give, in my opinion, we should not allow to deletion of admin users.
9. Create multiple users with the same name "test_user" and different auth sources. Run a delete-user command like "python setup.py delete-user test_user" It should delete all the users with the matching name, but it gives a "User not found" error. On providing "--auth-source" the user gets deleted.
10. Similar to the above when we run "python setup.py get-user test_user" it should give details of all the users with the name "test_user". On providing "--auth-source" it works.
11. On running "python setup.py get-users --auth-source ldap" I should be able to get all the ldap users only, currently command returns all the users.
12. "update-user" with --nonadmin not working. Try to update any admin user to nonadmin.
13. "update-external-user" have --role option instead of "--admin" and "--nonadmin", it should be consistent with "add-external-user".
14. We are not allowing to update passwords for external users from GUI, should we remove the "password" option from "update-external-user"?
15. "update-external-user" not working at all. I tried the following commands:

- python setup.py update-external-user akshay --auth-source ldap --email aj@edb.com
- python setup.py update-external-user akshay --email aj@edb.com (Multiple users with username akshay) 


16. "python setup.py get-prefs --id" returns nothing.
17. In the set-prefs Usage can we specify the PREF_OPTIONS should be provided in the form of key=value? It would be good for the users.
18. Is it possible to provide the datatype or possible values of preferences with the "get-prefs" command? It's very difficult for the user to know the datatype or possible value while setting the preferences.
19. Usage: setup.py dump-servers [OPTIONS] DUMP_SERVERS in this DUMP_SERVERS should be "OUTPUT_FILE". Similarly, LOADSERVERS should be changed to "INPUT_FILE".
20. We should add error handling if --sqlite-path is incorrect, previously it was not visible, but with CLI changes complete exception with traceback is visible. Can you please check all the other commands/options if possible?
21. --servers not working with dump-servers please check. Not giving any error, but servers are not dumped.
22. --replace not working with load-servers. Not giving any error, but servers are not loaded.

**Documentation Updates:**
1) Replace the occurrences of '--load-servers' with 'load-servers'.
2) Replace the occurrences of '--dump-servers' with 'dump-servers'.
3) Replace **--server** in this command "/path/to/python /path/to/setup.py --dump-servers output_file.json --server 1 2 5" to **--servers**
4) Provide an example in the documentation when dump-servers and load-servers are used with --user and --auth-source.

khushboovashi commented:
On Thu, Dec 7, 2023 at 7:58 PM Akshay Joshi ***@***.***>
wrote:

> ***@***.**** requested changes on this pull request.
>
>    1. Fixed SonarQube issues in the 'setup.py' file.
>
> *GUI Comments:*
>
>    1. Should we stop adding/deleting/updating user options in the Desktop
>    Mode?
>
>  I don't see any reason why we want to stop it.

>
>    1. Remove the "--auth-source" option from the "add-user" command as it
>    is of no use.
>
> Done

>
>    1. Remove auth-source as 'internal' from the "add-external-user"
>    command, as we will unable to add internal users using this command. Add a
>    validation if someone provides then the user should be warned.
>
> Done

>
>    1. Password validation of  6 characters should be there in the
>    "add-user", "update-user" command.
>
> Done

>
>    1. Do we need the "confirm password" option in the "add-user" command?
>
> We have an update-user, so no need to confirm password.

>
>    1. Please check the already existing email or username if possible.
>
> Done

>
>    1. What is the difference/use between/of --console and --no-console? I
>    have run both but have not found any difference, tried with add-user and
>    add-external-user.
>
> Removed for now.

>
>    1. delete-user is a very risky option to give, in my opinion, we
>    should not allow to deletion of admin users.
>
> Now, I added a confirmation message before deletion.

>
>    1. Create multiple users with the same name "test_user" and different
>    auth sources. Run a delete-user command like "python setup.py delete-user
>    test_user" It should delete all the users with the matching name, but it
>    gives a "User not found" error. On providing "--auth-source" the user gets
>    deleted.
>
> For delete users, Internal is the  default auth source, so if the user
will not provide auth source, it will check only for Internal users.
delete-user should not delete all the users with the same username
regardless of auth source.

>
>    1. Similar to the above when we run "python setup.py get-user
>    test_user" it should give details of all the users with the name
>    "test_user". On providing "--auth-source" it works.
>
> This seems reasonable to me, so implemented.

>
>    1. On running "python setup.py get-users --auth-source ldap" I should
>    be able to get all the ldap users only, currently command returns all the
>    users.
>
> This was not part of the implementation, now it's done.

>
>    1. "update-user" with --nonadmin not working. Try to update any admin
>    user to nonadmin.
>
> Done

>
>    1. "update-external-user" have --role option instead of "--admin" and
>    "--nonadmin", it should be consistent with "add-external-user".
>
> Done

>
>    1. We are not allowing to update passwords for external users from
>    GUI, should we remove the "password" option from "update-external-user"?
>
> It's a good observation, fixed.

>
>    1. "update-external-user" not working at all. I tried the following
>    commands:
>
>
>    - python setup.py update-external-user akshay --auth-source ldap
>    --email ***@***.***
>    - python setup.py update-external-user akshay --email ***@***.***
>    (Multiple users with username akshay)
>
> Fixed

>
>
>
>    1. "python setup.py get-prefs --id" returns nothing.
>
> Actually, it's for internal usage but it will display in help
unfortunately. I have not mentioned it in documentation.

>
>    1. In the set-prefs Usage can we specify the PREF_OPTIONS should be
>    provided in the form of key=value? It would be good for the users.
>
> We don't have those details as we do not store it in a database.

>
>    1. Is it possible to provide the datatype or possible values of
>    preferences with the "get-prefs" command? It's very difficult for the user
>    to know the datatype or possible value while setting the preferences.
>
> We don't have those details as we do not store it in a database. Users can
check the Preference dialog, where we have mappings for this on hover.


>
>    1. Usage: setup.py dump-servers [OPTIONS] DUMP_SERVERS in this
>    DUMP_SERVERS should be "OUTPUT_FILE". Similarly, LOADSERVERS should be
>    changed to "INPUT_FILE".
>
> Done

>
>    1. We should add error handling if --sqlite-path is incorrect,
>    previously it was not visible, but with CLI changes complete exception with
>    traceback is visible. Can you please check all the other commands/options
>    if possible?
>
> I have made some adjustments, which will display a shorter pretty
exception.

>
>    1. --servers not working with dump-servers please check. Not giving
>    any error, but servers are not dumped.
>
> Done

>
>    1. --replace not working with load-servers. Not giving any error, but
>    servers are not loaded.
>
> Done

>
>
> *Documentation Updates:*
>
>    1. Replace the occurrences of '--load-servers' with 'load-servers'.
>
> Done

>
>    1. Replace the occurrences of '--dump-servers' with 'dump-servers'.
>
> Done

>
>    1. Replace *--server* in this command "/path/to/python
>    /path/to/setup.py --dump-servers output_file.json --server 1 2 5" to
>    *--servers*
>
> Done

>
>    1. Provide an example in the documentation when dump-servers and
>    load-servers are used with --user and --auth-source.
>
> Done

>
>
> ------------------------------
>
> In web/setup.py
> <https://github.com/pgadmin-org/pgadmin4/pull/7031#discussion_r1416962478>
> :
>
> > +    webserver = WEBSERVER
> +
> +
> +class ManageUsers:
> +
> +    @app.command()
> +    def add_user(email: str, password: str,
> +                 role: Annotated[Optional[bool], typer.Option(
> +                     "--admin/--nonadmin")] = False,
> +                 auth_source: Optional[str] = INTERNAL,
> +                 active: Annotated[Optional[bool],
> +                                   typer.Option("--active/--inactive")] = True,
> +                 console: Optional[bool] = True,
> +                 json: Optional[bool] = False
> +                 ):
> +        """Add user. """
>
> Should be """ Add Internal User."
> ------------------------------
>
> In web/setup.py
> <https://github.com/pgadmin-org/pgadmin4/pull/7031#discussion_r1416965399>
> :
>
> > +        ManageUsers.create_user(data, console, json)
> +
> +    @app.command()
> +    def add_external_user(username: str,
> +                          auth_source: authType = authType.oauth2,
> +                          email: Optional[str] = None,
> +                          role: Annotated[Optional[bool],
> +                                          typer.Option(
> +                                              "--admin/--nonadmin")] = False,
> +                          active: Annotated[Optional[bool],
> +                                            typer.Option(
> +                                                "--active/--inactive")] = True,
> +                          console: Optional[bool] = True,
> +                          json: Optional[bool] = False
> +                          ):
> +        """Add user. """
>
> Should be """ Add External User like LDAP, Kerberos" or something similar.
> As it is visible on the command line when running setup.py with the
> '--help' option.
> ------------------------------
>
> In web/setup.py
> <https://github.com/pgadmin-org/pgadmin4/pull/7031#discussion_r1416967537>
> :
>
> > +                    print('User deleted successfully.')
> +                else:
> +                    print('Something went wrong. ' + str(msg))
> +
> +    @app.command()
> +    def update_user(email: str,
> +                    password: Optional[str] = None,
> +                    role: Annotated[Optional[bool],
> +                                    typer.Option("--admin/--nonadmin"
> +                                                 )] = False,
> +                    active: Annotated[Optional[bool],
> +                                      typer.Option("--active/--inactive"
> +                                                   )] = None,
> +                    console: Optional[bool] = True,
> +                    json: Optional[bool] = False
> +                    ):
>
> Add a docstring maybe """ Update internal user."""
> ------------------------------
>
> In web/setup.py
> <https://github.com/pgadmin-org/pgadmin4/pull/7031#discussion_r1416968444>
> :
>
> > +                status, msg = update_user(uid, data)
> +                if status:
> +                    _user = ManageUsers.get_users(username=email,
> +                                                  auth_source=INTERNAL,
> +                                                  console=False)
> +                    ManageUsers.display_user(_user[0], console, json)
> +                else:
> +                    print('Something went wrong. ' + str(msg))
> +
> +    @app.command()
> +    def get_users(username:Optional[str] = None,
> +                  auth_source: authType = authType.internal,
> +                  console:Optional[bool] = True,
> +                  json:Optional[bool] = False
> +                  ):
> +        app = create_app(config.APP_NAME + '-cli')
>
> Add a docstring maybe """ Get the list of users."""
> ------------------------------
>
> In web/setup.py
> <https://github.com/pgadmin-org/pgadmin4/pull/7031#discussion_r1416969427>
> :
>
> > +            if not console:
> +                return users_data
> +
> +    @app.command()
> +    def update_external_user(username: str,
> +                             auth_source: authType = authType.oauth2,
> +                             email: Optional[str] = None,
> +                             password: Optional[str] = None,
> +                             role: Optional[str] = None,
> +                             active: Annotated[
> +                                 Optional[bool],
> +                                 typer.Option("--active/--inactive")] = None,
> +                             console: Optional[bool] = True,
> +                             json: Optional[bool] = False
> +                             ):
> +        data = dict()
>
> Add a docstring maybe """ Update external user."""
> ------------------------------
>
> In web/setup.py
> <https://github.com/pgadmin-org/pgadmin4/pull/7031#discussion_r1416971879>
> :
>
> > +
> +class ManagePreferences:
> +
> +    def get_user(username=None, auth_source=INTERNAL):
> +        app = create_app(config.APP_NAME + '-cli')
> +        usr = None
> +        with app.test_request_context():
> +            usr = User.query.filter_by(username=username,
> +                                       auth_source=auth_source).first()
> +            if not usr:
> +                return None
> +            return usr.id
> +
> +    @app.command()
> +    def get_prefs(id: Optional[bool] = None, json: Optional[bool] = False):
> +        app = create_app(config.APP_NAME + '-cli')
>
> Add a docstring maybe """ Get all the preferences."""
> ------------------------------
>
> In web/setup.py
> <https://github.com/pgadmin-org/pgadmin4/pull/7031#discussion_r1416972696>
> :
>
> > +            if id:
> +                return all_preferences
> +            else:
> +                if json:
> +                    json_formatted_str = jsonlib.dumps(
> +                        {"Preferences": all_preferences},
> +                        indent=0)
> +                    console.print(json_formatted_str)
> +                else:
> +                    console.print(table)
> +
> +    @app.command()
> +    def set_prefs(username, pref_options: List[str],
> +                  auth_source: authType = authType.internal,
> +                  json: Optional[bool] = False):
> +        user_id = ManagePreferences.get_user(username, auth_source)
>
> Add a docstring maybe """ Set preferences."""
> ------------------------------
>
> In web/setup.py
> <https://github.com/pgadmin-org/pgadmin4/pull/7031#discussion_r1416974375>
> :
>
> > +def main():
> +    app()
> +
> +
> +if __name__ == "__main__":
> +    main()
>
> Add a comment on why we need this main() function.
> ------------------------------
>
> In docs/en_US/user_management.rst
> <https://github.com/pgadmin-org/pgadmin4/pull/7031#discussion_r1418640226>
> :
>
> > +    # to specify a role, 1 for admin user, 2 for non-admin user:
> +
> +    /path/to/python /path/to/setup.py add-user ***@***.*** password --role 1
>
> Correct the comment and command as there is no --role option in the
> "add-user" command.
> ------------------------------
>
> In docs/en_US/user_management.rst
> <https://github.com/pgadmin-org/pgadmin4/pull/7031#discussion_r1418641090>
> :
>
> > +
> +    /path/to/python /path/to/setup.py add-user ***@***.*** password
> +
> +    # to specify a role, 1 for admin user, 2 for non-admin user:
> +
> +    /path/to/python /path/to/setup.py add-user ***@***.*** password --role 1
> +
> +    # to specify user's status
> +
> +    /path/to/python /path/to/setup.py add-user ***@***.*** password --active
> +    /path/to/python /path/to/setup.py add-user ***@***.*** password --inactive
> +
> +Add External User
> +*****************
> +
> +To add external authentication user, invoke ``setup.py`` with ``add_external_user`` command line option,
>
> add_external_user should be add-external-user
> ------------------------------
>
> In docs/en_US/user_management.rst
> <https://github.com/pgadmin-org/pgadmin4/pull/7031#discussion_r1418651627>
> :
>
> > +    /path/to/python /path/to/setup.py add-external-user ***@***.*** ldap
> +
> +    # to specify an email:
> +
> +    /path/to/python /path/to/setup.py add-external-user ldapuser ldap --email ***@***.***
> +
> +    # to specify a role, 1 for admin user, 2 for non-admin user:
> +
> +    /path/to/python /path/to/setup.py add-external-user ldapuser ldap --role 1
> +
> +    # to specify user's status
> +
> +    /path/to/python /path/to/setup.py add-external-user ***@***.*** ldap --active
> +    /path/to/python /path/to/setup.py add-external-user ***@***.*** ldap --inactive
> +
> +Update User
> +***********
> +
> +To update user, invoke ``setup.py`` with ``update-user`` command line option, followed by
> +email address. password, role and active are updatable fields.
> +
> +.. code-block:: bash
> +
> +    /path/to/python /path/to/setup.py update-user ***@***.*** --password new-password
> +
> +    # to specify a role, 1 for admin user, 2 for non-admin user:
> +
> +    /path/to/python /path/to/setup.py update-user ***@***.*** password --role 1
> +
> +    # to specify user's status
> +
> +   /path/to/python /path/to/setup.py update-user ***@***.*** password --active
> +   /path/to/python /path/to/setup.py update-user ***@***.*** password --inactive
> +
> +Update External User
> +********************
> +
> +To update the external user, invoke ``setup.py`` with ``update-external-user`` command line option,
> +followed by username and auth source. email, password, role and active are updatable fields.
> +
> +.. code-block:: bash
> +
> +    # to change email address:
> +
> +    /path/to/python /path/to/setup.py update-external-user ldap ldapuser --email ***@***.***
> +
> +    # to change a role:
> +
> +    /path/to/python /path/to/setup.py update-user ldap ldapuser --role 2
> +
> +    # to change user's status
> +
> +   /path/to/python /path/to/setup.py update-user ldap ldapuser --active
> +   /path/to/python /path/to/setup.py update-user ldap ldapuser --inactive
> +
> +Delete User
> +***********
> +
> +To delete the user, invoke ``setup.py`` with ``delete-user`` command line option, followed by
> +username and auth_source. For Internal users, email adress will be used instead of username.
> +
> +.. code-block:: bash
> +
> +    /path/to/python /path/to/setup.py delete-user ***@***.*** internal
> +    /path/to/python /path/to/setup.py delete-user ldapuser  ldap
>
> Correct/Revisit all the commands:
>
>    - ldap can't be provided directly it should be with "--auth-source".
>
> Done

>
>    - there is no --role option in the "add-external-user" command.
>
> Done

>
>    - Check whether we can provide the options directly or if it needs the
>    key-value pair. Example --username is required before directly giving
>    username in the command.
>
>  There are 2 parts of the command, Arguments and Options. Arguments are
mandatory fields and Options are optional. So, whenever we have Options,
--options (--username)  will be there.

Thanks,
Khushboo

>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pgadmin-org/pgadmin4/pull/7031#pullrequestreview-1767046213>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADNOKSOLRH2NPY2GOAXDBQLYIHHBNAVCNFSM6AAAAABAFW6TA2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTONRXGA2DMMRRGM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

jkreet commented:
there is no typer in official docker image dpage/pgadmin4:8.4 
got an error ModuleNotFoundError: No module named 'typer'
